src/node_api_jsrt.cc

```diff
@@ -382,131 +555,127 @@ napi_status napi_define_class(napi_env e,
   JsPropertyIdRef configurableProperty;
   CHECK_JSRT(JsCreatePropertyIdUtf8("configurable", 12, &configurableProperty));
 
   JsPropertyIdRef enumerableProperty;
   CHECK_JSRT(JsCreatePropertyIdUtf8("enumerable", 10, &enumerableProperty));
 
-  for (int i = 0; i < property_count; i++) {
+  for (size_t i = 0; i < property_count; i++) {
     const napi_property_descriptor* p = properties + i;
 
     JsValueRef descriptor;
     CHECK_JSRT(JsCreateObject(&descriptor));
 
     JsValueRef configurable;
-    CHECK_JSRT(JsBoolToBoolean(!(p->attributes & napi_dont_delete), &configurable));
+    CHECK_JSRT(JsBoolToBoolean((p->attributes & napi_configurable), &configurable));
     CHECK_JSRT(JsSetProperty(descriptor, configurableProperty, configurable, true));
 
     JsValueRef enumerable;
-    CHECK_JSRT(JsBoolToBoolean(!(p->attributes & napi_dont_enum), &enumerable));
+    CHECK_JSRT(JsBoolToBoolean((p->attributes & napi_enumerable), &enumerable));
     CHECK_JSRT(JsSetProperty(descriptor, enumerableProperty, enumerable, true));
 
-    if (p->method) {
-      JsPropertyIdRef valueProperty;
-      CHECK_JSRT(JsCreatePropertyIdUtf8("value", 5, &valueProperty));
-      JsValueRef method;
-      CHECK_NAPI(napi_create_function(e, p->method, p->data,
-        reinterpret_cast<napi_value*>(&method)));
-      CHECK_JSRT(JsSetProperty(descriptor, valueProperty, method, true));
-    }
-    else if (p->getter || p->setter) {
-      if (p->getter) {
+    if (p->getter != nullptr || p->setter != nullptr) {
+      napi_value property_name;
+      CHECK_NAPI(jsrtimpl::JsNameValueFromPropertyDescriptor(p, &property_name));
+
+      if (p->getter != nullptr) {
         JsPropertyIdRef getProperty;
         CHECK_JSRT(JsCreatePropertyIdUtf8("get", 3, &getProperty));
         JsValueRef getter;
-        CHECK_NAPI(napi_create_function(e, p->getter, p->data,
-          reinterpret_cast<napi_value*>(&getter)));
+        CHECK_NAPI(napi_create_property_function(env, property_name,
+          p->getter, p->data, reinterpret_cast<napi_value*>(&getter)));
         CHECK_JSRT(JsSetProperty(descriptor, getProperty, getter, true));
       }
 
-      if (p->setter) {
+      if (p->setter != nullptr) {
         JsPropertyIdRef setProperty;
         CHECK_JSRT(JsCreatePropertyIdUtf8("set", 5, &setProperty));
         JsValueRef setter;
-        CHECK_NAPI(napi_create_function(e, p->setter, p->data,
-          reinterpret_cast<napi_value*>(&setter)));
+        CHECK_NAPI(napi_create_property_function(env, property_name,
+          p->setter, p->data, reinterpret_cast<napi_value*>(&setter)));
         CHECK_JSRT(JsSetProperty(descriptor, setProperty, setter, true));
       }
-    }
-    else {
+    } else if (p->method != nullptr) {
+      napi_value property_name;
+      CHECK_NAPI(jsrtimpl::JsNameValueFromPropertyDescriptor(p, &property_name));
+
+      JsPropertyIdRef valueProperty;
+      CHECK_JSRT(JsCreatePropertyIdUtf8("value", 5, &valueProperty));
```

Comment:
This is being created in a loop- in the chakrashim, there is a perf optimization to cache some of the more frequently used propertyidrefs- we should have a todo to do the same in napi-jsrt